plugins/flatpak/gs-flatpak.c

```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	g_object_unref (helper->cancellable);
+	g_clear_error (&helper->error);
+	g_free (helper);
+}
+
+static GsFlatpakCopyProcessHelper *
+gs_flatpak_copy_process_helper_new (GsFlatpak *self,
+				    GsApp *app,
+				    GCancellable *cancellable,
+				    gulong cancelled_id)
+{
+	GsFlatpakCopyProcessHelper *helper = g_new0 (GsFlatpakCopyProcessHelper,
+						     1);
+	helper->self = self;
+	helper->app = g_object_ref (app);
+	helper->cancellable = g_object_ref (cancellable);
+	helper->cancelled_id = cancelled_id;
+	helper->finished = FALSE;
+	helper->error = NULL;
+	return helper;
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(GsFlatpakCopyProcessHelper, gs_flatpak_copy_process_helper_free)
+
+static void
+copy_process_watch_cb (GPid pid, gint status, gpointer user_data)
+{
+	GsFlatpakCopyProcessHelper *helper =
+		(GsFlatpakCopyProcessHelper*) user_data;
+	g_autoptr(GError) error = NULL;
+
+	if (!g_cancellable_is_cancelled (helper->cancellable) && status) {
+		g_set_error (&helper->error,
+			     GS_PLUGIN_ERROR,
+			     GS_PLUGIN_ERROR_FAILED,
+			     "failed to copy flatpak app to removable "
+			     "media: `flatpak create-usb` failed with "
+			     "status %d", status);
+	}
+
+	g_cancellable_disconnect (helper->cancellable, helper->cancelled_id);
+	g_spawn_close_pid (pid);
+
+	helper->finished = TRUE;
+
+	/* once the copy terminates (successfully or not), set plugin status to
+	 * update UI accordingly */
+	gs_plugin_status_update (helper->self->plugin, helper->app,
+				 GS_PLUGIN_STATUS_FINISHED);
+
+	// update metadata to ensure the USB category includes this new app and
+	// we can immediately download an app, copy to USB, delete from
+	// computer, and install from USB without leaving the app page
+	gs_flatpak_refresh (helper->self, 0, GS_PLUGIN_REFRESH_FLAGS_METADATA,
+			    NULL, &error);
+	if (error != NULL) {
+		g_warning ("failed to refresh flatpak metadata after copying "
+			"app to USB: %s", error->message);
+	}
+}
+
+static void
+app_copy_cancelled_cb (GCancellable *cancellable,
+		       gpointer user_data)
+{
+	GPid pid = GPOINTER_TO_INT (user_data);
+
+	// terminate the process which is copying the app
+	kill (pid, SIGTERM);
+}
+
+gboolean
+gs_flatpak_app_copy (GsFlatpak *self,
+		     GsApp *app,
+		     const gchar *dest,
+		     GCancellable *cancellable,
+		     GError **error)
+{
+	// this is used in an async function but we block here until that
+	// returns so we won't auto-free while other threads depend on this
+	g_autoptr (GsFlatpakCopyProcessHelper) helper = NULL;
+	const gchar *app_id;
+	const gchar *app_arch;
+	const gchar *app_branch;
+	g_autofree gchar* app_ref = NULL;
+	gboolean spawn_retval;
+	g_autofree GStrv argv;
+	GPid child_pid;
+	gulong cancelled_id;
+	const gchar* app_collection_id;
+
+	app_id = gs_flatpak_app_get_ref_name (app);
+	app_arch = gs_flatpak_app_get_ref_arch (app);
+	app_branch = gs_flatpak_app_get_ref_branch (app);
+	app_ref = g_strdup_printf ("%s/%s/%s", app_id,
```

Comment:
Nitpick: Maybe set all these together with their declaration?

Reply:
I thought it was convention to have no assignments (except for `NULL`) at the beginning of a block but maybe I was confusing `-Wdeclaration-after-statement` (which doesn't mind assignments mixed into the declaration section after all).

Fixed locally.